packages/mui-styled-engine/src/StyledEngineProvider/StyledEngineProvider.js

```diff
@@ -37,32 +54,67 @@ if (typeof document === 'object') {
       head.prepend(insertionPoint);
     }
   }
-  /**
-   * This is for client-side apps only.
-   * A custom sheet is required to make the GlobalStyles API injected above the insertion point.
-   * This is because the [sheet](https://github.com/emotion-js/emotion/blob/main/packages/react/src/global.js#L94-L99) does not consume the options.
-   */
-  class MyStyleSheet extends StyleSheet {
-    insert(rule, options) {
-      if (this.key && this.key.endsWith('global')) {
-        this.before = insertionPoint;
+}
+
+function getCache(injectFirst, enableCssLayer) {
+  if (injectFirst || enableCssLayer) {
+    /**
+     * This is for client-side apps only.
+     * A custom sheet is required to make the GlobalStyles API injected above the insertion point.
+     * This is because the [sheet](https://github.com/emotion-js/emotion/blob/main/packages/react/src/global.js#L94-L99) does not consume the options.
+     */
+    class MyStyleSheet extends StyleSheet {
+      insert(rule, options) {
+        if (privateForTest.insert) {
+          return privateForTest.insert(rule, options);
+        }
+        if (this.key && this.key.endsWith('global')) {
+          this.before = insertionPoint;
+        }
+        return super.insert(rule, options);
       }
-      return super.insert(rule, options);
     }
+    const emotionCache = createEmotionCache(
+      {
+        key: 'css',
+        insertionPoint: injectFirst ? insertionPoint : undefined,
+      },
+      MyStyleSheet,
+    );
+    if (enableCssLayer) {
+      const prevInsert = emotionCache.insert;
+      emotionCache.insert = (...args) => {
+        if (!args[1].styles.startsWith('@layer')) {
+          // avoid nested @layer
+          args[1].styles = `@layer mui {${args[1].styles}}`;
+        }
+        return prevInsert(...args);
+      };
+    }
```

Comment:
Could we define this in the `insert` definition above, instead of accessing and setting `emotionCache.insert`? I think this would make the code cleaner.

Reply:
That won't work. The `cache.insert` is a different function to `insert` in the StyleSheet class.

We need to intercept the cache because the serialized styles has to be modified before going through [stylis](https://github.com/emotion-js/emotion/blob/main/packages/cache/src/index.ts#L162).

By moving the logic to StyleSheet insert, it won't go through stylis and it might not work as expected. Another reason is the cache.insert contain source map logic. It we modify the StyleSheet insert, I believe it will make the source map incorrect.

Reply:
Thanks for the explanation, I didn't realize it was different `insert`s 😅